asynciterator.ts

```diff
@@ -568,6 +567,47 @@ export class AsyncIterator<T> extends EventEmitter {
   clone(): ClonedIterator<T> {
     return new ClonedIterator<T>(this);
   }
+
+  /**
+   * An AsyncIterator is async iterable.
+   * This allows iterators to be used via the for-await syntax.
+   * @returns {ESAsyncIterator<T>} An EcmaScript AsyncIterator
+   */
+  [Symbol.asyncIterator](): ESAsyncIterator<T> {
+    const it = this;
+    // An EcmaScript AsyncIterator exposes the next() function that can be invoked repeatedly
+    return {
+      next(): Promise<IteratorResult<T>> {
+        return new Promise<IteratorResult<T>>((resolve, reject) => {
+          it.on('error', reject);
+          function tryResolve(): void {
+            const value = it.read();
+            if (value !== null) {
+              // Immediately resolve if we have read an element
+              it.removeListener('error', reject);
+              resolve({ done: false, value });
+            }
+            else if (it.done) {
+              // Close if we're done
+              it.removeListener('error', reject);
+              resolve({ done: true, value: undefined });
+            }
+            else {
+              // In all other cases, wait for the iterator to become readable or ended
+              const nextCallback = () => {
+                it.removeListener('readable', nextCallback);
+                it.removeListener('end', nextCallback);
+                tryResolve();
+              };
```

Comment:
You should be able to define this outside of `tryResolve` rather than re-defining it every time.

Reply:
Done!